<div style="border:solid Chocolate 2px; padding: 40px">

Привет, меня зовут Александр Куимов. Я буду ревьюером твоего проекта. Ты можешь обращаться ко мне на "ты"😏 Надеюсь, тебя также не смутит, если я буду обращаться к тебе на "ты", но если это неудобно, обязательно скажи об этом!

**Пожалуйста, не перемещай, не изменяй и не удаляй  мои комментарии**, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 

Прошу реагировать на все мои комментарии, по которым ты вносишь правки в проект, своими комментариями по шаблону, показанному чуть ниже. На каждое мое замечание нужно отвечать комментарием студента в маркдаун ячейке **ниже моей. Это нужно, чтобы не создавалась путаница**.😉

Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным.
</div>

<br/>
 
<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
    
<br/>    
<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b>👋:</b> В такой цветовой ячейке - образце я прошу тебя оставлять твои комментарии. **Обязательно добавь заголовок "Комментарий студента" к каждой твоей ячейке**. Если исправляешь проект во второй раз и далее, не забывай пожалуйста указывать номер итерации, например, "Комментарий студента v.2". У меня к этому моменту уже появится нумерация в моих комментариях.
</div> 

<br/>    
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самому найти и исправить ее. На реальной работе твой руководитель будет поступать также, и я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку!🤓
    
 

# Проект спринта 7 "Секреты Темнолесья" 

- Автор: Маликов Виталий
- Дата: 08-04-2025

### Цели и задачи проекта

<font color='#777778'>Перечислите цель проекта и задачи, которые вы решаете. Для этого можно использовать описание проекта, но полезнее будет сформулировать основную цель проекта самостоятельно.</font>

Привлечь новую аудиторию и подготовить статью о развитии индустрии игр в начале XXI века. Выпустить статью хотят на одном известном IT-ресурсе. По идее коллег статья должна привлечь внимание людей, которые любят старые игры, и заодно заинтересовать их «Секретами Темнолесья».

В статье-исследовании хотят сделать обзор игровых платформ, изучить объёмы продаж игр разных жанров и региональные предпочтения игроков. Акцент в статье хотят сделать на играх жанра RPG — так называют компьютерные ролевые игры, в которых игроки управляют персонажами.


Познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных.
Перед анализом вам необходимо дополнительно сделать следующее:

Отобрать данные по времени выхода игры. Вам нужен период с 2000 по 2013 год включительно.

Категоризовать игры по оценкам пользователей и экспертов. Выделите три категории:

    высокая оценка — с оценкой от 8 до 10 и от 80 до 100, включая правые границы интервалов.
    
    средняя оценка — с оценкой от 3 до 8 и от 30 до 80, не включая правые границы интервалов.
    
    низкая оценка — с оценкой от 0 до 3 и от 0 до 30, не включая правые границы интервалов.
    
Выделить топ-7 платформ по количеству игр, выпущенных за весь требуемый период.

### Описание данных

<font color='#777778'>Здесь приведите описание данных.</font>

Предстоит работать с историческими данными, которые коллеги собрали из открытых источников. В них есть информация о продажах игр, сделанных в разных жанрах и выпущенных на разных платформах, а также пользовательские и экспертные оценки игр.

Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:

Name — название игры.

Platform — название платформы.

Year of Release — год выпуска игры.

Genre — жанр игры.

NA sales — продажи в Северной Америке (в миллионах проданных копий).

EU sales — продажи в Европе (в миллионах проданных копий).

JP sales — продажи в Японии (в миллионах проданных копий).

Other sales — продажи в других странах (в миллионах проданных копий).

Critic Score — оценка критиков (от 0 до 100).

User Score — оценка пользователей (от 0 до 10).

Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

<font color='#777778'>Перечислите основные шаги проекта или напишите оглавление. Используйте описание проекта, чтобы зафиксировать основные шаги.</font>

---

#### Загрузка и знакомство с данными
Загрузите необходимые библиотеки Python и данные датасета /datasets/new_games.csv.

Познакомьтесь с данными: выведите первые строки и результат метода info().

Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.

Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

#### Проверка ошибок в данных и их предобработка
Это самая объёмная и важная часть проекта. Теперь вам нужно внимательно изучить данные и провести их предобработку. Что нужно проверить:
1. Названия, или метки, столбцов датафрейма:
Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.

Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания('_') 

2. Типы данных:
Если встречаются некорректные типы данных, предположите их причины.

При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу int64. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.
В числовых столбцах могут встретиться строковые значения, например,unknown или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

3. Наличие пропусков в данных:
Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.

Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.

Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.

Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.

Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

4. Явные и неявные дубликаты в данных:
Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.

При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.
Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

В процессе подготовки данных вы могли удалять данные, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях. После проведения предобработки данных напишите общий промежуточный вывод.

#### Фильтрация данных
Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например df_actual.

#### Категоризация данных
Проведите категоризацию данных:
Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

 <div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
<b>Все отлично!👍:</b> 
    
Вижу твое добавленное название проекта, описание данных. Молодец! Это поможет тебе расставлять акценты в выводах

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b>
    

Не стоит оставлять код ячейки только с текстовой информацией. Для ячеек только с текстовой информацией принято использовать Markdown-формат ячеек.
        
Насчет этих ячеек - их лучше просто удалить        

<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b>Сделано!</b> 
</div> 

<div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [1]:
import pandas as pd

df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [2]:
df.head()

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b>
    

Стоит использовать  метод `head()` вместо вывода df при проведении первичного аналитическог осмотра

<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b>Сделано!</b> 
</div> 

<div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

<font color='#777778'>Используйте ячейки типа Markdown для промежуточных выводов и расширенных комментариев к действиям с данными. </font>

Таблица с данными содержит 16956 строк и 11 столбцов, 4 из которых имеют тип float64, а 7 - object. Объем составляет более 1.4 Мб. В 6 столбцах (Name, Year of Release, Genre, Critic Score, User Score, Rating) есть пропуски.

Столбец EU sales, JP sales, User Score имеют тип object, который необходимо поменять на float64.

Столбец Year of Release имеет тип данных float64 - необходимо изменить на int64.

Также необходимо привесте все столбцы к стилю snake case, а названия сделать в нижнем регистре.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 
    
С данными познакомились, проведен первичный аналитический осмотр!) Молодец, что детально изучаешь каждый столбец и указываешь основные ошибки
    
Полученная информация поможет нам в дальнейшем с предобработкой данных
</div>

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [3]:
# Выводим на экран названия всех столбцов датафрейма
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [4]:
# Приводим все столбцы к стилю snake case
columns_snake_case = [
    'name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
    'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'  
]

df.columns = columns_snake_case

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 
    
Отлично, названия столбцов приведены к стилю snake_case, это делает датафрейм более удобным для работы и чтения кода 
    
В целом тут можно использовать цепочку методов, добавлю другие альтернативные способы переименования названия столбцов:      
        
- чтобы переименовать колонки, можно также использовать метод `rename()`



- также интересным является подход с использованием list comprehensions:
        
`df.columns = [x.lower().replace(' ', '_') for x in df.columns.values]`
    
    
Данный подход использует list comprehensions, что может быть быстрее на больших датасетах благодаря оптимизации Python для list comprehensions. Этот метод также предоставляет гибкость в добавлении дополнительных операций или условий при необходимости. Однако, он может быть менее очевидным для тех, кто только начинает работать с Python. Этот подход "интересен" тем, что он показывает, как одна и та же задача может быть решена разными способами, и некоторые из них могут быть более эффективными или гибкими в зависимости от конкретного контекста   
    
            

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

#### Возможные причины некорректных типов данных
- Столбцы EU sales, JP sales, User Score имеют тип object, который необходимо поменять на float64. 
- Столбец Year of Release имеет тип данных float64 - необходимо изменить на int64.

Неправильный тип данных сформировался,возможно, из-за ввода значений вручную. Таким образом могут сожержаться символы/объекты/смешанные данные, которые нельзя чётко отнести к числовым, временным или категориальным.
Кроме того, наличие побелов также может стать причиной типа данных 'object'

In [5]:
df.dtypes

name                object
platform            object
year_of_release    float64
genre               object
na_sales           float64
eu_sales            object
jp_sales            object
other_sales        float64
critic_score       float64
user_score          object
rating              object
dtype: object

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [6]:
print('Количество пропусков в столбцах')
print(df.isna().sum())
print('Количество пропусков в столбцах в относительных значениях')
print(df.isna().sum() / df.shape[0] * 100)

Количество пропусков в столбцах
name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8714
user_score         6804
rating             6871
dtype: int64
Количество пропусков в столбцах в относительных значениях
name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.000000
jp_sales            0.000000
other_sales         0.000000
critic_score       51.391838
user_score         40.127389
rating             40.522529
dtype: float64


- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


- В столбце year_of_release 275 пропусков (примерно 1,6%). Можно удалить, т.к. года неизвестны, а нам необходимо  выбрать игры с 2000 по 2013 год
- В name и genre по 2 пропуска (0,01%). Можно удалить. 
- В critic_score - 8714 (более 51%). 
- В user_score (6804) и rating (6871), что составляет примерно по 40%

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Проверить колличество пропусков в % можно еще вот так: 
    
    
    pd.DataFrame(round(df.isna().mean()*100,)).style.background_gradient('coolwarm')
    
- Можно также использовать f-строку для вывода процентов и округления данных :.2%   https://python-scripts.com/f-strings

In [7]:
# Удаляем строки из датасета, в которых отсутствует год, имя и жанр
df = df.dropna(subset=['year_of_release', 'name'])

In [8]:
# В critic_score есть пропуски, но все они числового типа (float64). Заменим пропуски на -1
df['critic_score'] = df['critic_score'].fillna(-1)

In [9]:
# В user_score есть и пропуски и другие объекты. Приводим к float64 и заменим пропуски на -1
df['user_score'] = pd.to_numeric(df['user_score'], errors = 'coerce', downcast = 'float')
df['user_score'] = df['user_score'].fillna(-1)

In [10]:

# Приводим eu_sales и jp_sales к float64. В этих столбцах нет пропусков, но, вероятно есть другие объекты. 
# Приводим такие столбцы к числовому типу данных, заменив строковые значения на пропуски.
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors = 'coerce', downcast = 'float')
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors = 'coerce', downcast = 'float')

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 
    
Использование pd.to_numeric() с параметром errors='coerce' для обработки некорректных значений — это правильное решение, которое позволяет избежать ошибок при преобразовании 
    
<div class="alert alert-warning">

    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Перед проведением обработки типов данных желательно посмотреть на уникальные значения в столбцах eu_sales, jp_sales и user_score, далее найти причины некорректных типов. Затем уже можно принимать решение об обработке    

In [11]:
df['user_score'].unique()

array([ 8. , -1. ,  8.3,  8.5,  6.6,  8.4,  8.6,  7.7,  6.3,  7.4,  8.2,
        9. ,  7.9,  8.1,  8.7,  7.1,  3.4,  5.3,  4.8,  3.2,  8.9,  6.4,
        7.8,  7.5,  2.6,  7.2,  9.2,  7. ,  7.3,  4.3,  7.6,  5.7,  5. ,
        9.1,  6.5,  8.8,  6.9,  9.4,  6.8,  6.1,  6.7,  5.4,  4. ,  4.9,
        4.5,  9.3,  6.2,  4.2,  6. ,  3.7,  4.1,  5.8,  5.6,  5.5,  4.4,
        4.6,  5.9,  3.9,  3.1,  2.9,  5.2,  3.3,  4.7,  5.1,  3.5,  2.5,
        1.9,  3. ,  2.7,  2.2,  2. ,  9.5,  2.1,  3.6,  2.8,  1.8,  3.8,
        0. ,  1.6,  9.6,  2.4,  1.7,  1.1,  0.3,  1.5,  0.7,  1.2,  2.3,
        0.5,  1.3,  0.2,  0.6,  1.4,  0.9,  1. ,  9.7], dtype=float32)

In [12]:
# В eu_sales есть пропуски. Заменим их на -1
df['eu_sales'] = df['eu_sales'].fillna(-1)

In [13]:
# В jp_sales есть пропуски. Заменим их на -1
df['jp_sales'] = df['jp_sales'].fillna(-1)

<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b>Вопрос имею: </b> 
- Если честно, то неочень понял, что не так со столбцом user_score
</div> 

<div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>
    
вижу, что сейчас с ним все хорошо)  Замена пропусков на среднее или медиану  могла бы исказить результаты - использование заглушки - предпочтительней

In [14]:
# Приводим year_of_release к int64 для удобства
df['year_of_release'] = df['year_of_release'].astype('int64')

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера  <a class="tocSkip"></h2>
    
<b>На доработку🤔:</b>

    
На этом шаге предполагается обработать пропуски в абсолютно всех признаках с наличием пропусков, дать обоснования используемых подходов к обработке пропусков

In [15]:
# Заменим пропуски в столбце rating на "NO"
df['rating'] = df['rating'].fillna('NO')

<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b>Заменил значения NaN в rating на NO. В других столбцах пропусков не нашел  </b> 
</div> 

<div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [16]:
# Переведем в нижний регистр
df['name'] = df['name'].str.lower()
df['genre'] = df['genre'].str.lower()
df['rating'] = df['rating'].str.upper()

In [17]:
# Выведем количество уникальных значений в некоторых столбцах
unique_values = df[['name', 'year_of_release', 'platform', 'genre', 'rating']].nunique()
unique_values

name               11426
year_of_release       37
platform              31
genre                 12
rating                 9
dtype: int64

In [18]:
# Считаем количество дубликатов
duplicated_rows = df.duplicated().sum()

duplicated_rows

235

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [19]:
# Удаляем явные дибликаты
df_cleaned = df.drop_duplicates()

df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16444 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16444 non-null  object 
 1   platform         16444 non-null  object 
 2   year_of_release  16444 non-null  int64  
 3   genre            16444 non-null  object 
 4   na_sales         16444 non-null  float64
 5   eu_sales         16444 non-null  float32
 6   jp_sales         16444 non-null  float32
 7   other_sales      16444 non-null  float64
 8   critic_score     16444 non-null  float64
 9   user_score       16444 non-null  float32
 10  rating           16444 non-null  object 
dtypes: float32(3), float64(3), int64(1), object(4)
memory usage: 1.3+ MB


- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

Найдено 235 явных дубликатов при помощи метода .duplicated().sum() 

На следующем шаге удалили выявленные дубликаты при помощи .drop_duplicated() во всем датафрейме (subset = None), оставив первое значение (keep='first') и при это не изменив начальный датасет (inplace = False)

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [20]:
display(16956 - df_cleaned.shape[0], df_cleaned.shape[0]/ 16956 * 100)

512

96.98041991035622

- После проведения предобработки данных напишите общий промежуточный вывод.

В датасете осталось 16444 строки, что составляет 97% от начального датасета. В ходе предобработки данных было удалено 512 строк.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Методы  `.duplicated()` и `.drop_duplicates()` использованы корректно)
    
Отлично, что подсчитан процент/доля отфильтрованных данных  

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [21]:
# Получаем датасет, в котором только игры с годом выпуска 2000 - 2013
df_actual = df_cleaned.loc[(2000 <= df_cleaned['year_of_release']) & (df_cleaned['year_of_release'] <= 2013)]

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Фильтрация данных проведена корректно)  Можно еще посмотреть, сколько значений было удалено

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [22]:
df_actual['user_score_group'] = pd.cut(df['user_score'], bins = [0, 3, 8, 10], labels = ['низкая оценка', ' средняя оценка', 'высокая оценка'])

/tmp/ipykernel_114/851701691.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_actual['user_score_group'] = pd.cut(df['user_score'], bins = [0, 3, 8, 10], labels = ['низкая оценка', ' средняя оценка', 'высокая оценка'])


<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b>Еще здесь вопрос есть. Почему в верхней и нижней строках вываливается ошибка, хотя при этом продолжает считать все правильно? </b> 
</div> 

<div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>
    
Отличный вопрос — ты поймал один из самых частых "вроде работает, но Pandas ругается" кейсов.

---

Что происходит?

Твоя строка:

```python
df_actual['user_score_group'] = pd.cut(df['user_score'], bins=[0, 3, 8, 10], labels=['низкая оценка', ' средняя оценка', 'высокая оценка'])
```

вызывает предупреждение (а не ошибку!) **`SettingWithCopyWarning`**, потому что `df_actual` — скорее всего **вырезка (slice)** из `df`, и Pandas **не уверен**, работает ли ты с **копией** или **представлением** оригинального DataFrame.

---

Почему результат считается правильно?
Да, результат действительно создаётся и выглядит как будто всё работает. Но Pandas предупреждает:  
> "ты, возможно, редактируешь временную копию, и твои изменения могут не сохраниться как ты ожидаешь".

Это может проявиться в более сложных кейсах — например, в цепочках фильтрации или при использовании `.loc`.

---

Как это исправить?
Явно укажи, что хочешь модифицировать `df_actual` через `.loc`:

```python
    
df_actual.loc[:, 'user_score_group'] = pd.cut(df_actual['user_score'], bins=[0, 3, 8, 10],
                                              labels=['низкая оценка', ' средняя оценка', 'высокая оценка'])
```

И ещё важно: используй колонку из **того же DataFrame**, а не `df['user_score']`, если работаешь с `df_actual`.

---

Итог:
- **Предупреждение**, а не ошибка — можно игнорировать, но лучше поправить.
- Pandas беспокоится о "неявных копиях", чтобы ты случайно не редактировал что-то, что не сохранится.
- Используй `.loc` и следи, чтобы вся операция шла через один и тот же DataFrame

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [23]:
df_actual['critic_score_group'] = pd.cut(df['critic_score'], bins = [0, 29, 79, 100], labels = ['низкая оценка', ' средняя оценка', 'высокая оценка'])

/tmp/ipykernel_114/3111052648.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_actual['critic_score_group'] = pd.cut(df['critic_score'], bins = [0, 29, 79, 100], labels = ['низкая оценка', ' средняя оценка', 'высокая оценка'])


In [24]:
df_actual

name platform  \
0                                            wii sports      Wii   
2                                        mario kart wii      Wii   
3                                     wii sports resort      Wii   
6                                 new super mario bros.       DS   
7                                              wii play      Wii   
...                                                 ...      ...   
16947                     men in black ii: alien escape       GC   
16949                woody woodpecker in crazy castle 5      GBA   
16950  score international baja 1000: the official game      PS2   
16952                                  lma manager 2007     X360   
16954                                  spirits & spells      GBA   

       year_of_release     genre  na_sales   eu_sales  jp_sales  other_sales  \
0                 2006    sports     41.36  28.959999      3.77         8.45   
2                 2008    racing     15.68  12.760000      3.79         3.29   
3                 2009    sports     15.61  10.930000      3.28         2.95   
6                 2006  platform     11.28   9.140000      6.50         2.88   
7                 2006      misc     13.96   9.180000      2.93         2.84   
...                ...       ...       ...        ...       ...          ...   
16947             2003   shooter      0.01   0.000000      0.00         0.00   
16949             2002  platform      0.01   0.000000      0.00         0.00   
16950             2008    racing      0.00   0.000000      0.00         0.00   
16952             2006    sports      0.00   0.010000      0.00         0.00   
16954             2003  platform      0.01   0.000000      0.00         0.00   

       critic_score  user_score rating user_score_group critic_score_group  
0              76.0         8.0      E   средняя оценка     средняя оценка  
2              82.0         8.3      E   высокая оценка     высокая оценка  
3              80.0         8.0      E   средняя оценка     высокая оценка  
6              89.0         8.5      E   высокая оценка     высокая оценка  
7              58.0         6.6      E   средняя оценка     средняя оценка  
...             ...         ...    ...              ...                ...  
16947          -1.0        -1.0      T              NaN                NaN  
16949          -1.0        -1.0     NO              NaN                NaN  
16950          -1.0        -1.0     NO              NaN                NaN  
16952          -1.0        -1.0     NO              NaN                NaN  
16954          -1.0        -1.0     NO              NaN                NaN  

[12781 rows x 13 columns]

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [25]:
# Группируем данные и считаем количество игр в выделенных категориях (user_score_group)
grouped_data_user = df_actual.groupby('user_score_group', as_index=False)['name'].count()
grouped_data_user

user_score_group  name
0    низкая оценка   132
1   средняя оценка  4317
2   высокая оценка  2033

In [26]:
# Группируем данные и считаем количество игр в выделенных категориях (critic_score_group)
grouped_data_critic = df_actual.groupby('critic_score_group', as_index=False)['name'].count()
grouped_data_critic

critic_score_group  name
0      низкая оценка    55
1     средняя оценка  5422
2     высокая оценка  1692

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера  <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b> 

Нужно получить две таблички, в которых будут отражены только 3 категории (а не 9): высокая оценка, средняя оценка, низкая оценка. Дополнительно дробить каждую из этих категорий на численные подкатегории не нужно. Прошу исправить этот момент   

<div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [27]:
top_platforms = df_actual.groupby('platform', as_index=False).size().sort_values(by='size', ascending=False)
top_platforms.head(7)

platform  size
9       PS2  2127
2        DS  2120
15      Wii  1275
12      PSP  1180
17     X360  1121
10      PS3  1087
4       GBA   811

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Отлично, что выведен ТОП-7 платформ по количеству игр)

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

Подключили библиотеку Pandas (import pandas)

Загрузили данные, используя pd.read_csv, и изучили их

Исходная таблица содержала 16956 строк и 11 столбцов с разными типами данных. Объем составлял более 1.4 Мб. В 6 столбцах были пропуски.

Далее перешли к предобработке данных:
- Привели все столбцы к стилю snake case, заменив старые названия колонок на новые
- Изучили данные на пропуски, после чего удалили 275 строк с пропусками года, т.к. по заданию они дальше не участвовали бы решении задач
- Перевели ряд столюцов в нижний и верхние регистры (.upper()/.lower())
- Привели числовые столбцы к соответствующему типу (.to_numeric()) с указанием оптимальной разрядности для чисел
- Определили, что в данных 235 явных дубликатов (.duplicated()), от которых избавились (.drop_duplicates())
- Всего в ходе очистки данных удалили 512 строк (примерно 3%). Таки образом в датасете осталось 16444 строки

Далее перешли к фильтрации:
- Отобрали только те игры, которые были выпущены в период с 2000 по 2013 год (.loc)

Далее провели категоризацию данных:
- Разделили игры на 3 категории в зависимости от оценок критиков и пользователей (.cut())
- Сгруппировали по новым столбцам 'user_score_group' и 'critic_score_group' (.groupby()). Получили, что чаще всего мнения критиков и пользователей сходились в "средней" оценке игр 
- Выделили ТОП-7 платформ по количеству игр в рассматриваемый период (.sort_values()). Самой популярной оказалась PS2, на которой вышло 2127 игр. На втором и третьем месте DS и Wii с результатом 2120 и 1275 игр соответственно

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Молодец, что не забываешь написать итоговый вывод по проекту. Добавлю, хорошей практикой является написание выводов по предобработке — используй пожалуйста ее в следующих проектах, когда будешь готовить данные 

</div>   
    
    
<div style="border:solid Chocolate 2px; padding: 40px">


<h2> Итоговый комментарий ревьюера (общий вывод по проекту)  <a class="tocSkip"> </h2>    
    
  
У меня сложилось хорошее впечатление о проекте. Молодец! Осмысленная аналитика и дельная работа - многое удалось как надо)
    
Отмечу отдельные положительные моменты проекта🙂:
    
- Была проделана значительная работа по обработке и анализу данных! 
- Твой вывод чётко отражает проделанные шаги, что важно для демонстрации результатов исследования.   
    
Есть несколько моментов всего, на которые стоит ещё раз взглянуть, я указал их в моих комментариях по ходу проекта.
    
Предлагаю тебе доработать проект по моим комментариям, чтобы довести его до совершенства.   
    

 
</div>


<div style="border:solid Chocolate 2px; padding: 40px">


<h2>  Итоговый комментарий ревьюера 2 (итоговый вывод по проекту) <a class="tocSkip"> </h2>    
    
      
  
Теперь почти идеально, молодец! Принимаю твой проект)
    
Могу порекомендовать тебе отличную книгу про Python:
    
https://monster-book.com/avtomatizaciya-zadach-s-python    
    
По визуализации могу посоветовать следующие материалы:
    
1. https://github.com/esokolov/ml-course-hse/blob/master/2022-fall/seminars/sem02-charts.ipynb
    
2. https://github.com/Palladain/Deep_Python/blob/main/Lectures/Lecture_3/Lecture_3_Python.ipynb    
    
Советую тебе обратить внимание на библиотеки `sweetviz` и `ydata_profiling`, которые помогут в проведении более тщательного EDA анализа. Данные библиотеки хороши для максимизации комфорта презентации результатов анализа бизнес-пользователям. 
    
Очень просты в использоовании, на вход кладется датафрейм: `ydata_profiling.ProfileReport(df)`
    
Цель нашего проекта достигнута. Рад был тебе помочь)
    
Удачи тебе в следующем проекте!
    
**Буду очень благодарен, если ты поставишь мне положительную оценку за проект в форме на странице курса, это важно для меня. Достаточно поставить оценку, а комментарий можешь написать по желанию:)**     
    
</div>    
